hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -159,6 +161,151 @@ def trivial(self):
+            return self.end(i - 1)
+
+    def end(self, i):
+        """Equivalent to self[i].end."""
+        return self.__endpoints[i]
+
+    def bounds(self, i):
+        """Equivalent to self[i].bounds."""
+        return (self.start(i), self.end(i))
+
+    def all_bounds(self):
+        """Equivalent to [(b.start, b.end) for b in self]."""
+        result = []
+        prev = 0
+        for e in self.__endpoints:
+            result.append((prev, e))
+            prev = e
+        return result
+
+    def __len__(self):
+        return len(self.__endpoints)
+
+    def __known_block(self, i):
+        try:
+            return self.__blocks[i]
+        except (KeyError, IndexError):
+            return None
+
+    def __getitem__(self, i):
+        n = len(self)
+        if i < -n or i >= n:
+            raise IndexError("Index %d out of range [-%d, %d)" % (i, n, n))
+        if i < 0:
+            i += n
+        assert i >= 0
+        result = self.__known_block(i)
+        if result is not None:
+            return result
+
+        # We store the blocks as a sparse dict mapping indices to the
+        # actual result, but this isn't the best representation once we
+        # stop being sparse and want to use most of the blocks. Switch
+        # over to a list at that point.
+        if self.__sparse and len(self.__blocks) * 2 >= len(self):
+            new_blocks = [None] * len(self)
+            for k, v in self.__blocks.items():
+                new_blocks[k] = v
+            self.__sparse = False
+            self.__blocks = new_blocks
+            assert self.__blocks[i] is None
+
+        start = self.start(i)
+        end = self.end(i)
+
+        # We keep track of the number of blocks that have actually been
+        # instantiated so that when every block that could be instantiated
+        # has been we know that the list is complete and can throw away
+        # some data that we no longer need.
+        self.__count += 1
+
+        # Integrity check: We can't have allocated more blocks than we have
+        # positions for blocks.
+        assert self.__count <= len(self)
+        result = Block(
+            start=start,
+            end=end,
+            index=i,
+            forced=start in self.owner.forced_indices,
+            all_zero=not any(self.owner.buffer[start:end]),
+        )
+        try:
+            self.__blocks[i] = result
+        except IndexError:
+            assert isinstance(self.__blocks, list)
+            assert len(self.__blocks) < len(self)
+            self.__blocks.extend([None] * (len(self) - len(self.__blocks)))
+            self.__blocks[i] = result
+
+        # If this happens then we have now fully calculated every block
+        # and don't need to keep the reference to the owner around. We delete
+        # it here so that there is no circular reference to make the gc work
+        # harder.
+        if self.__count == len(self) and isinstance(self.owner, ConjectureResult):
```

Comment:
This can potentially be missed if the entire block list is populated before ownership is handed over.

I don’t know how likely that is in practice, and either way the penalty is only a missed GC optimization, so it’s not a huge deal.

Reply:
Hmm. That's a good point, and easy to fix. I'll factor this out into a method and call it from `transfer_ownership` as well.